src/AppInstallerRepositoryCore/Microsoft/Schema/1_4/DependenciesTable.cpp

```diff
@@ -0,0 +1,375 @@
+    {
+        SQLite::Savepoint savepoint = SQLite::Savepoint::Create(connection, std::string{ s_DependenciesTable_Table_Name } + "add_dependencies_v1_4");
+
+        auto dependencies = GetDependencies(manifest, Manifest::DependencyType::Package);
+        if (!dependencies.size())
+        {
+            return;
+        }
+
+        InsertManifestDependencies(connection, manifestRowId, dependencies);
+
+        savepoint.Commit();
+    }
+
+    bool DependenciesTable::UpdateDependencies(SQLite::Connection& connection, const Manifest::Manifest& manifest, SQLite::rowid_t manifestRowId)
+    {
+        SQLite::Savepoint savepoint = SQLite::Savepoint::Create(connection, std::string{ s_DependenciesTable_Table_Name } + "update_dependencies_v1_4");
+
+        const auto dependencies = GetDependencies(manifest, Manifest::DependencyType::Package);
+        auto existingDependencies = GetDependenciesByManifestRowId(connection, manifestRowId);
+
+        // Get dependencies to add.
+        std::set<Manifest::Dependency> toAddDependencies;
+        std::copy_if(
+            dependencies.begin(),
+            dependencies.end(),
+            std::inserter(toAddDependencies, toAddDependencies.begin()),
+            [&](Manifest::Dependency dep)
+            {
+                auto depRow = GetPackageRowIdAndVersion(connection, dep);
+                return existingDependencies.find(depRow.value()) == existingDependencies.end();
+            }
+        );
+
+        // Get dependencies to remove.
+        std::vector<SQLite::rowid_t> toRemoveDependencies;
+        std::vector<Manifest::Dependency> missingPackageNodes;
+        std::set<std::pair<SQLite::rowid_t, Utility::NormalizedString>> dependenciesPackageIdAndVersion;
+        std::for_each(
+            dependencies.begin(),
+            dependencies.end(),
+            [&](auto& dep)
+            {
+                auto depRow = GetPackageRowIdAndVersion(connection, dep);
+                if (!depRow.has_value())
+                {
+                    missingPackageNodes.emplace_back(dep);
+                    return;
+                }
+
+                dependenciesPackageIdAndVersion.emplace(depRow.value());
+            }
+        );
+
+        ThrowOnMissingPackageNodes(missingPackageNodes);
+
+        std::for_each(
+            existingDependencies.begin(),
+            existingDependencies.end(),
+            [&](std::pair<std::pair<SQLite::rowid_t, Utility::NormalizedString>, SQLite::rowid_t> row)
+            {
+                if (dependenciesPackageIdAndVersion.find(row.first) == dependenciesPackageIdAndVersion.end())
+                {
+                    toRemoveDependencies.emplace_back(row.second);
+                }
+            }
+        );
+
+        InsertManifestDependencies(connection, manifestRowId, toAddDependencies);
+        RemoveDependenciesByRowIds(connection, toRemoveDependencies);
```

Comment:
>RemoveDependenciesByRowIds [](http://example.com/codeflow?start=8&length=26)

Given that the PK should be { manifest, package id }, you could just as easily remove by that instead of forcing the use of the rowid here.  Then you wouldn't need the public API to return the unnecessary dependency rowid.